### PR TITLE
Feature : create a new component to display tree views

### PR DESCRIPTION
**Pull request by SirineMhedhbi**
### Context
See https://github.com/ontoportal-lirmm/bioportal_web_ui/issues/293

### Changes

- Create a new component with turbo frame 
-  Replace the current code of Jquery simple tree with this new component tree view
- Remove the code of Jquery simple tree

image

### CODE REVIEW

File: app/javascript/controllers/simple_tree_controller.js [hunk 1]
```diff
@@ -1,76 +0,0 @@
...
-
-    connect() {
-        this.simpleTreeCollection = useSimpleTree(this.element,
-            this.#afterClick.bind(this),
-            this.#afterAjaxError.bind(this),
-            this.#beforeAjax.bind(this)
-        )
-
-
-        this.simpleTreeCollection.ready(() => {
```
[the remark this answers is not in the document]
Comment (SirineMhedhbi) [reply app/javascript/controllers/simple_tree_controller.js:19]: remove only use simpleTree

File: app/views/ontologies/concepts_browsers/_concepts_tree.html.haml [hunk 2]
```diff
@@ -16,6 +16,4 @@
                         'chosen-enable-colors-value': 'true'}}
   -# Class tree
   %div#sd_content.card.p-1.py-3{style: 'overflow-y: scroll; height: 60vh;'}
-    = render TurboFrameComponent.new(id: 'concepts_tree_view',
-                                     src: "/ajax/classes/treeview?ontology=#{@ontology.acronym}&conceptid=#{escape(@concept.id)}&concept_schemes=#{params[:concept_schemes]}&auto_click=false&language=#{request_lang}",
```
Comment (SirineMhedhbi) [inline app/views/ontologies/concepts_browsers/_concepts_tree.html.haml:20]: think to remove the action behind

File: app/views/ontologies/concepts_browsers/_concepts_tree.html.haml [hunk 3]
```diff
@@ -16,6 +16,4 @@
...
-                                     data: {'turbo-frame-target': 'frame'})
\ No newline at end of file
+    = render TreeViewComponent.new([], @ontology, @concept.id,  params[:concept_schemes], "EN", params, data: {'turbo-frame-target': 'frame'})
```
Comment (SirineMhedhbi) [inline app/views/ontologies/concepts_browsers/_concepts_tree.html.haml:19]: put the request language instead of "EN" and see why it doesn't work if so.

Comment (SirineMhedhbi) [reply app/views/ontologies/concepts_browsers/_concepts_tree.html.haml:19]: remove the first [] argument

File: app/helpers/application_helper.rb [hunk 4]
```diff
@@ -155,6 +155,89 @@ def draw_note_tree_leaves(notes,level,output,key)
...
+        end
+        @root = LinkedData::Client::Models::Class.new(read_only: true)
+        @root.children = rootNode.sort{|x,y| (x.prefLabel || "").downcase <=> (y.prefLabel || "").downcase}
+      end
+    end
+
+    @concept
+  end
+
   def draw_tree(root, id = nil, concept_schemes = [])
```
Comment (SirineMhedhbi) [inline app/helpers/application_helper.rb:241]: think of how to migrate this ruby code to view component and haml

File: app/views/concepts/_child_nodes.html.haml [hunk 5]
```diff
@@ -1,7 +1,8 @@
-- output =""
-- for child in @children
-  - output << tree_link_to_concept(child: child, ontology_acronym: child.explore.ontology.acronym, active_style: '', node: @concept)
-  - if child.hasChildren
-    - output << tree_link_to_children(child: child, concept_schemes: @schemes)
-  - output << "</li>"
-= raw output
+= turbo_frame_tag child_id(@concept)+'_childs' do
```
Comment (SirineMhedhbi) [inline app/views/concepts/_child_nodes.html.haml:1]: think of how to remove this file because it's duplicated with draw tree helper

File: app/views/ontologies/_treeview.html.haml [hunk 6]
```diff
@@ -1,10 +0,0 @@
-= turbo_frame_tag 'concepts_tree_view' do
-  #tree_wrapper.hide-if-loading
-    %ul.simpleTree{data:{controller: 'simple-tree',
```
Comment (SirineMhedhbi) [inline app/views/ontologies/_treeview.html.haml:3]: create controller js to copy the autoclick part

File: app/components/tree_view_component.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+class TreeViewComponent < ViewComponent::Base
+   include ApplicationHelper
+   include ActionView::Helpers::TagHelper
+    
+    def initialize(tree_data, ontology, conceptid, concept_schemes, language, params, data: )
+      @tree_data = tree_data
```
Comment (syphax-bouazzouni) [inline app/components/tree_view_component.rb:8]: those variables are not used, `@tree_data`; `@language`; `@data`; `@params`

File: app/components/tree_view_component/tree_view_component.html.haml [hunk 8]
```diff
@@ -0,0 +1,20 @@
+%turbo-frame{ data: { controller: 'tree-view'}, "data-turbo-frame-target": 'frame', id: "tree-frame" }
+  .tree_wrapper.hide-if-loading
+    %ul.simpleTree
+      %li.root
+        %ul
+          = draw_tree(@root, @concept.id, (params[:concept_schemes] || '').split(','))
```
Comment (syphax-bouazzouni) [inline app/components/tree_view_component/tree_view_component.html.haml:6]: extract draw_tree into a view component

File: app/controllers/concepts_controller.rb [hunk 9]
```diff
@@ -3,6 +3,9 @@
 class ConceptsController < ApplicationController
   include MappingsHelper
   include ConceptsHelper
+  include TurboHelper
+  include ApplicationHelper
```
Comment (syphax-bouazzouni) [inline app/controllers/concepts_controller.rb:7]: try to remove `include ApplicationHelper` if possible

File: app/helpers/application_helper.rb [hunk 10]
```diff
@@ -155,6 +155,89 @@ def draw_note_tree_leaves(notes,level,output,key)
     end
   end
 
+  def get_class(params)
+
+    lang = "EN"
```
Comment (syphax-bouazzouni) [inline app/helpers/application_helper.rb:160]: remove lang = "EN"

**Review by syphax-bouazzouni — CHANGES_REQUESTED** [review]

**Comment by syphax-bouazzouni** [comment]
@SirineMhedhbi what is your progress on this?

File: app/components/tree_view_component.rb [hunk 11]
```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: false
+
+class TreeViewComponent < ViewComponent::Base
+   include ApplicationHelper
+include ActionView::Helpers::TagHelper
+
```
Comment (syphax-bouazzouni) [inline app/components/tree_view_component.rb:6]: try to remove 
```
include ApplicationHelper
include ActionView::Helpers::TagHelper
```

this component is really important, and need to be independent the most possible of any external behaviors (external dependecies).

File: app/javascript/controllers/index.js [hunk 12]
```diff
@@ -72,3 +71,9 @@ application.register("turbo-frame", TurboFrameController)
 
 import TurboFrameErrorController from "./turbo_frame_error_controller"
 application.register("turbo-frame-error", TurboFrameErrorController)
+
+import ToggleController from "./toggle_controller"
```
Comment (syphax-bouazzouni) [inline app/javascript/controllers/index.js:75]: remove this

File: app/javascript/controllers/toggle_controller.js [hunk 13]
```diff
@@ -0,0 +1,20 @@
+// app/javascript/controllers/toggle_controller.js
+import {Controller} from "@hotwired/stimulus"
+
+export default class extends Controller {
```
Comment (syphax-bouazzouni) [inline app/javascript/controllers/toggle_controller.js:4]: remove this file, it is no more used

File: app/views/concepts/_child_nodes.html.haml (hunk 5 above)
Comment (syphax-bouazzouni) [reply app/views/concepts/_child_nodes.html.haml:1]: Still to do;

Here the idea will be to create a component called for example `TreeChildren`

File: app/views/ontologies/concepts_browsers/_concepts_tree.html.haml (hunk 2 above)
Comment (syphax-bouazzouni) [reply app/views/ontologies/concepts_browsers/_concepts_tree.html.haml:20]: Still to do;

The thing to here is to clean the code that was used before:
* in routes.rb, remove  `/ajax/classes/treeview`
* in concept_controller.rb, remove  the action `show_tree`

but before doing that, you need to be sure that it is no longer used anywhere.

**Review by syphax-bouazzouni — CHANGES_REQUESTED** [review]
Things to add / fix:
* See the comments in " files changed "
* The icon color is not good 
image
* Try to change the icon style, to something better, for example: https://www.svgrepo.com/svg/507824/plus-square and https://www.svgrepo.com/svg/524752/minus-square
*  Add the lines of each nodes as before:
image
*  Try to add some padding/margin between the elements
image

File: app/components/tree_view_component.rb [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+      return string if node.children.nil? || node.children.empty?
+  
+      node.children.sort! { |a, b| (a.prefLabel || a.id).downcase <=> (b.prefLabel || b.id).downcase }
+      node.children.each do |child|
+        active_style = child.id.eql?(id) ? "active" : ''
+  
+        # This fake root will be present at the root of "flat" ontologies, we need to keep the id intact
+  
+        if child.id.eql?('bp_fake_root')
+          string << tree_link_to_concept(child: child, ontology_acronym: '',
```
Comment (syphax-bouazzouni) [inline app/components/tree_view_component.rb:36]: this part of the code and the one in the file _child_ndes is similar, try to re-use the same code, either same helper or same component

File: app/helpers/application_helper.rb [hunk 15]
```diff
@@ -155,69 +155,124 @@ def draw_note_tree_leaves(notes,level,output,key)
     end
   end
 
-  def draw_tree(root, id = nil, concept_schemes = [])
```
Comment (syphax-bouazzouni) [inline app/helpers/application_helper.rb:158]: you removed `draw_tree`, here but it is still used in the file `_treeview`

Comment (syphax-bouazzouni) [reply app/helpers/application_helper.rb:158]: May be the solution would be to just change in _treeview, draw_tree by your tree component

**Comment by syphax-bouazzouni** [comment]
## Changes done on my side 
* A new UI to the tree link
image
* Migrate the Concept tree (Tree, list, and date views), Properties tree, Schemes tree, and Collection to use the new TreeComponent
* Removed old js code (the files `bp_property_tree.js`, `useSimpleTree.js`,  `jquery.history.js` and `jquery.simple.tree.js.erb` were removed)
* Add tests for all the trees

**Review by syphax-bouazzouni — APPROVED** [review]

**Comment by jonquet** [comment]
OK for release in v2.7.3
